### PR TITLE
[WIP] Combine label added and removed events

### LINKED ISSUE #1321: Join/group event messages together
I recall this being mentioned ages ago but we couldn't do it because the way (for example) added/removed label messages were constructed - But now they're just attributed labels we could start grouping them similar to GH.com

"BasThomas added `project-mgmt` `starter-task` and removed `accessibility` labels 18 hours ago"

### PR DESCRIPTION
**Pull request by SD10**
Resolves #1321 

~**This is extremely bug ridden**. The events, UI, and everything is inaccurate.~
Getting better but needs refinement.
Just showing some of the complexity involved. This is the first time I could get the project to build 😓

### GitHawk:
screen shot 2018-02-22 at 3 12 34 am

### GitHub:
screen shot 2018-02-22 at 3 14 51 am

### CODE REVIEW

File: Classes/Issues/Labeled/IssueLabeledModel.swift [hunk 1]
```diff
@@ -51,47 +55,100 @@ final class IssueLabeledModel: ListDiffable {
...
+                        .foregroundColor: labelColor.textOverlayColor ?? .black,
+                        .baselineOffset: 1, // offset for better rounded background colors
+                        MarkdownAttribute.label: LabelRepoDetails(owner: repoOwner, repo: repoName, label: title)
+                    ]
+                ))
+                attributedString.append(spaceString)
+            }
+        }
+
+        let lastDate = added.last?.date ?? removed.last?.date ?? Date()
```
Comment (SD10) [inline Classes/Issues/Labeled/IssueLabeledModel.swift]: Need to choose the greatest of the two dates here

File: Classes/Issues/Issue+IssueType.swift [hunk 2]
```diff
@@ -93,32 +93,56 @@ extension IssueOrPullRequestQuery.Data.Repository.IssueOrPullRequest.AsIssue: Is
...
+                        date: date,
+                        type: .added)
+                    var removed = [IssueLabeledModel.LabelDetails]()
+                    var added = [IssueLabeledModel.LabelDetails]()
+
+                    if let lastModel = results.last as? IssueLabeledModel,
+                        lastModel.actor == labeled.actor?.login {
+                        added = lastModel.added
+                        removed = lastModel.removed
+                        removed.append(newDetails)
```
Comment (SD10) [inline Classes/Issues/Issue+IssueType.swift]: This should be `added`

File: Classes/Issues/Issue+IssueType.swift [hunk 3]
```diff
@@ -93,32 +93,56 @@ extension IssueOrPullRequestQuery.Data.Repository.IssueOrPullRequest.AsIssue: Is
...
+                        date: date,
+                        type: .removed)
+                    var removed = [IssueLabeledModel.LabelDetails]()
+                    var added = [IssueLabeledModel.LabelDetails]()
+
+                    if let lastModel = results.last as? IssueLabeledModel,
+                        lastModel.actor == unlabeled.actor?.login {
+                        added = lastModel.added
+                        removed = lastModel.removed
+                        removed.append(newDetails)
```
Comment (SD10) [inline Classes/Issues/Issue+IssueType.swift]: `newDetails` only gets appended in this block, which is why it breaks for events with a single label

File: Classes/Issues/Issue+IssueType.swift (hunk 2 above)
Comment (SD10) [inline Classes/Issues/Issue+IssueType.swift]: `newDetails` only gets appended in this block, which is why it breaks for events with a single label

File: Classes/Issues/GithubClient+Issues.swift [hunk 4]
```diff
@@ -337,41 +337,44 @@ extension GithubClient {
...
         let newLabelNames = Set<String>(labels.map { $0.name })
 
-        var newEvents = [IssueLabeledModel]()
+        var addedDetails = [IssueLabeledModel.LabelDetails]()
+        var removedDetails = [IssueLabeledModel.LabelDetails]()
+
         for newLabel in labels {
             if !oldLabelNames.contains(newLabel.name) {
-                newEvents.append(IssueLabeledModel(
+                addedDetails.append(IssueLabeledModel.LabelDetails(
```
Comment (SD10) [inline Classes/Issues/GithubClient+Issues.swift]: Need to look into if I broke mutations with this change

File: Classes/Issues/Labeled/IssueLabeledModel.swift [hunk 5]
```diff
@@ -51,47 +55,100 @@ final class IssueLabeledModel: ListDiffable {
...
+                ))
+                attributedString.append(spaceString)
+            }
+        }
+
+        let lastDate = min(added.last?.date ?? Date(), removed.last?.date ?? Date())
 
         attributedString.append(NSAttributedString(
-            string: "  \(date.agoString)",
+            string: " labels \(lastDate.agoString)",
```
Comment (SD10) [inline Classes/Issues/Labeled/IssueLabeledModel.swift]: Needs to be `label(s)`

File: Classes/Issues/Issue+IssueType.swift (hunk 2 above)
Comment (rnystromtest) [reply Classes/Issues/Issue+IssueType.swift]: Test

Comment (rnystrom) [reply Classes/Issues/Issue+IssueType.swift]: test

**Comment by SD10** [comment]
Yeah, sorry about that. Let's close it. Trying to flatten the GitHub event timeline was too messy for this to be merged. If someone else wants to try it 👌

**Pull request closed without merging**